src/core/providers/vpc/qgsvirtualpointcloudprovider.cpp

```diff
@@ -0,0 +1,555 @@
+      }
+      else if ( geom.at( "type" ) == "MultiPolygon" )
+      {
+        for ( auto &geomPart : geom.at( "coordinates" ) )
+        {
+          QgsPolygonXY part;
+          for ( auto &geomRing : geomPart )
+          {
+            QgsPolylineXY ring;
+            for ( auto &coords : geomRing )
+            {
+              ring.append( QgsPointXY( coords.at( 0 ), coords.at( 1 ) ) );
+            }
+            part.append( ring );
+          }
+          geometry.append( part );
+        }
+      }
+      else
+      {
+        QgsDebugMsg( QStringLiteral( "Unexpected geometry type: %1" ).arg( QString::fromStdString( geom.at( "type" ) ) ) );
+      }
+      si.geometry = QgsGeometry::fromMultiPolygonXY( geometry );
+    }
+    catch ( std::exception &e )
+    {
+      QgsDebugMsg( QStringLiteral( "Malformed geometry item: %1" ).arg( QString::fromStdString( e.what() ) ) );
+      return;
+    }
+
+    if ( si.uri.startsWith( QLatin1String( "./" ) ) )
+    {
+      // resolve relative path
+      si.uri = fInfo.absoluteDir().absoluteFilePath( si.uri );
+    }
+
+    if ( f["properties"].contains( "pc:schemas" ) )
+    {
+      for ( auto &schemaItem : f["properties"]["pc:schemas"] )
+      {
+        attributeNames.insert( QString::fromStdString( schemaItem["name"] ) );
+      }
+    }
+
+    if ( transform.isValid() && !transform.isShortCircuited() )
+      try
+      {
+        si.geometry.transform( transform );
+      }
+      catch ( QgsCsException & )
+      {
+        QgsDebugMsg( QStringLiteral( "Cannot transform geometry to layer crs." ) );
+        continue;
+      }
+    mPolygonBounds->addPart( si.geometry );
+    mSubLayers.push_back( si );
+    mPointCount += si.count;
+  }
+  mExtent = mPolygonBounds->boundingBox();
+  populateAttributeCollection( attributeNames );
+}
+
+QgsGeometry QgsVirtualPointCloudProvider::polygonBounds() const
+{
+  return *mPolygonBounds;
+}
+
+void QgsVirtualPointCloudProvider::loadIndex( int i )
+{
+  QGIS_PROTECT_QOBJECT_THREAD_ACCESS
+
+  if ( i >= mSubLayers.size() || i < 0 )
+    return;
+
+  QgsPointCloudSubIndex &sl = mSubLayers[ i ];
+  // Index already loaded -> no need to load
+  if ( sl.index )
+    return;
+
+  if ( sl.uri.startsWith( QStringLiteral( "http" ), Qt::CaseSensitivity::CaseInsensitive ) )
+  {
+    if ( sl.uri.endsWith( QStringLiteral( "copc.laz" ), Qt::CaseSensitivity::CaseInsensitive ) )
+      sl.index.reset( new QgsRemoteCopcPointCloudIndex() );
+    else if ( sl.uri.endsWith( QStringLiteral( "ept.json" ), Qt::CaseSensitivity::CaseInsensitive ) )
+      sl.index.reset( new QgsRemoteEptPointCloudIndex() );
+  }
+  else
+  {
+    if ( sl.uri.endsWith( QStringLiteral( "copc.laz" ), Qt::CaseSensitivity::CaseInsensitive ) )
+      sl.index.reset( new QgsCopcPointCloudIndex() );
+    else if ( sl.uri.endsWith( QStringLiteral( "ept.json" ), Qt::CaseSensitivity::CaseInsensitive ) )
+      sl.index.reset( new QgsEptPointCloudIndex() );
+  }
+
+  if ( !sl.index )
+    return;
```

Comment:
worth writing a warning that such file was not recognized

Reply:
This warning will be spammed every time its `loadIndex()` is called. We could add the warning when creating the `QgsPointCloudSubIndex` object instead